商品購入機能実装

3so opened the pull request:
#What
出品された商品をクレジットカード決済によって購入できる機能の実装

#Why
フリマアプリを作成する上で、製品を購入する機能が必要であるため。

#機能確認用GIF動画
※製品購入ページのURLは「http://localhost:3000/items/【itemのid】/purchases」です

■ログインしていないユーザーは購入ページに遷移しようとすると、ログインページに遷移すること
　https://gyazo.com/e5424ac58fe759e819d46210698feb82

■出品者はURLを直接入力して購入ページに遷移しようとすると、トップページに遷移すること
　https://gyazo.com/5b4389b4e13d0aeb707c1b47be561ff1

■URLを直接入力して購入済み商品の購入ページへ遷移しようとすると、トップページに遷移すること
　https://gyazo.com/49d43b62ff2852595bef672de955376c

■クレジットカード決済ができること
　https://gyazo.com/8f7b6e0110aef8671804c6ea2675002c
　

3so committed:
Purchaseモデル、ShippingAddressモデルの作成

3so committed:
マイグレーションの実行

3so committed:
purchasesコントローラーの作成

3so committed:
purchasesコントローラーのindexアクションのルーティング設定

itemsコントローラーにネストさせる

3so committed:
purchasesコントローラーのindexアクション定義

3so committed:
purchases#indexのビュー設置

3so committed:
商品詳細ページから購入ページへのリンク設定

3so committed:
製品購入ページでの製品情報表示

3so committed:
Revert "製品購入ページでの製品情報表示"

This reverts commit eb89ac268bbe1aba143c8c59038bbcb5075787f4.

3so committed:
製品購入ページでの製品情報表示

3so committed:
PurchaseモデルとShippingAddressモデルのアソシエーション記述

3so committed:
turbolinksに関する記述の削除

3so committed:
情報を保存するカラム名の設定

3so committed:
purchase#createのルーティング設定

3so committed:
PurchaseShippingAddressモデルの作成

Formオブジェクトパターンを用いて、purchasesテーブルとshipping_addressテーブルに同時に情報を保存するため。

3so committed:
purchaseコントローラーのcreateアクション定義

3so committed:
purchase#indexのビュー編集

・PurchaseShippingAddressモデルのインスタンスに情報を保存するよう変更
・エラーメッセージを表示するよう設定

3so committed:
payjp.jsの読み込み

3so committed:
purchase#indexのビュー編集

・クレジットカード情報を入力するフォームにname属性を追加

3so committed:
PAY.jpによる決済を行うためのJavaScript作成

3so committed:
トークンの情報を送信できるよう編集

・PurchaseShippingAddressモデルを編集、token属性のゲッターとセッターを追加
・purchasesコントローラーを編集、PurchaseShippingAddressモデルにtokenの値を追加するよう変更

3so committed:
webpacker.rbの作成

・card.jsで環境変数を使用できるようにするため

3so committed:
payjpの導入

3so committed:
shipping_addressテーブルの修正

・phone_numberカラムが抜けていたので追加

3so committed:
purchasesコントローラーのcreateアクション編集

・payjpのgemを用いて、決済を行えるよう編集

3so committed:
token属性のバリデーション設定

3so committed:
Itemモデル、UserモデルにPurchaseモデルとのアソシエーション追加

3so committed:
itemsコントローラーのindexアクション、showアクション編集

・purchasesテーブルに保存している購入情報を送信するよう変更
・Sold Outの表示に必要なため
・indexアクションはn+1問題を考慮しincludesメソッドを仕様

3so committed:
商品一覧の編集

・商品が売れていればsold outを表示するよう変更

3so committed:
商品詳細ページの変更

・購入済みの商品にはsold outが出るよう変更
・購入済みの商品は編集・削除・購入ボタンが表示されないよう変更

3so committed:
purchasesコントローラーの編集

・未ログインユーザーが購入ページにアクセスしようとするとログイン画面に遷移させるよう変更
・出品者が購入ページにアクセスしようとするとトップページへ遷移するよう変更
・購入済み製品の購入ページにアクセスしようとするとトップページへ遷移するよう変更

3so committed:
PurchaseShippingAddressモデルの単体テスト実装

kokikurahashi commented on app/controllers/items_controller.rb:
```diff
@@ -22,6 +22,7 @@ def create
   end
 
   def show
+    @purchase = Purchase.find_by(item_id: @item.id)
   end
```
こちらは@item.purchaseとアソシエーションを用いれば良いため変数化の必要はないです！

3so replied:
変数化の必要がないため、記述を削除しました。

kokikurahashi commented on app/views/items/show.html.erb:
```diff
@@ -9,13 +9,11 @@
-      </div>
-      <%# //商品が売れている場合は、sold outを表示しましょう %>
-      <%# //商品購入機能未実装のため、非表示にしています %>
+      <% if @purchase != nil %>
+        <div class='sold-out'>
+          <span>Sold Out!!</span>
+        </div>
+      <% end %>
```
@item.purchaseを使用しましょう！

3so replied:
@item.purchaseを用いる形に記述し直しました。

kokikurahashi commented on app/views/items/show.html.erb:
```diff
@@ -27,17 +25,15 @@
+        <p class='or-text'>or</p>
+        <%= link_to '削除', item_path(@item.id), method: :delete, class:'item-destroy' %>
 
-    <% elsif user_signed_in? %>
-      <%# 商品購入機能が未実装のため、常に表示するようにしています %>
-      <%# 商品が売れていない場合はこちらを表示しましょう %>
-      <%= link_to '購入画面に進む', "#" ,class:"item-red-btn"%>
-      <%# //商品が売れていない場合はこちらを表示しましょう %>
-      <%# //商品購入機能が未実装のため、常に表示するようにしています %>
+      <% elsif user_signed_in? %>
+        <%= link_to '購入画面に進む', item_purchases_path(@item.id) ,class:"item-red-btn"%>
+      <% end %>
```
if文にもelsif文にもuser_signed_in? があるため一つ上のif文にまとめましょう！
`<% if @purchase == nil && user_signed_in? %>`

3so replied:
ご指摘いただいた形にて、ひとつ上のif文にまとめました。

kokikurahashi commented on spec/models/purchase_shipping_address_spec.rb:
```diff
@@ -0,0 +1,81 @@
+
+    context '配送先情報の保存' do
+
+      it 'すべての値が正しく入力されていれば保存できること' do
+        expect(@purchase_shipping_address).to be_valid
+      end
```
建物番号がなくても購入できるテストを追加しましょう！

3so replied:
建物番号がなくても購入できるテストを追加・実行し、問題ないことを確認しました。

kokikurahashi reviewed (commented):
レビュー依頼ありがとうございます！
コメントしましたので、ご確認をお願いいたします！

3so committed:
showアクションにおけるPurchaseモデルのインスタンス削除

・itemモデルのアソシエーションを利用することで生成の必要がなくなるため

3so committed:
items#showのビュー編集

・purchaseモデルのインスタンスを用いていた箇所を、itemモデルとpurchaseモデルのアソシエーションを利用する形に変更
・重複していたif文の条件をまとめた

3so committed:
PurchaseShippingAddressモデルの単体テスト変更

・「建物番号がなくても問題ないこと」を追加

3so commented:
ご確認・コメント頂き、ありがとうございます。
ご指摘いただいた箇所を修正いたしましたので、お手数をおかけいたしますが、今一度ご確認お願いいたします。

kokikurahashi commented on app/views/items/show.html.erb:
```diff
@@ -27,17 +25,14 @@
+        <%= link_to '商品の編集', edit_item_path(@item.id), method: :get, class: "item-red-btn" %>
+        <p class='or-text'>or</p>
+        <%= link_to '削除', item_path(@item.id), method: :delete, class:'item-destroy' %>
+      <% elsif %>
+        <%= link_to '購入画面に進む', item_purchases_path(@item.id) ,class:"item-red-btn"%>
+      <% end %>
```
elsifではなくelseにしましょう！（elseにすれば条件式は出品者!=購入者となるため）

3so replied:
条件式が空欄のelsifとなってしまっていた箇所をelseに修正しました。

kokikurahashi reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認させていただきました！
一点修正点についてコメントしましたので、ご確認をお願いいたします！

3so committed:
items#showのビュー編集

・条件式を指定しないelsifになってしまっていた部分をelseに修正

3so commented:
修正点をご確認いただき、ありがとうございます。
ご指摘いただいた点を再度修正いたしましたので、お手数をおかけいたしますがご確認お願いいたします。

kokikurahashi commented:
レビュー依頼ありがとうございます！
修正点問題なしです！
LGTMとさせていただきます！実装お疲れ様でした！
引き続き学習の方頑張っていきましょう！応援しております！
